server/src/future.rs

```diff
@@ -157,3 +158,42 @@ impl Stream for IntervalStream {
 		}
 	}
 }
+
+#[derive(Debug, Clone)]
+pub(crate) struct SessionClose(tokio::sync::broadcast::Sender<()>);
+
+impl SessionClose {
+	pub(crate) fn close(self) {
+		let _ = self.0.send(());
+	}
+
+	pub(crate) fn closed(&self) -> SessionClosedFuture {
+		SessionClosedFuture(BroadcastStream::new(self.0.subscribe()))
+	}
+}
+
+/// A future that resolves when the connection has been closed.
+#[derive(Debug)]
+pub struct SessionClosedFuture(BroadcastStream<()>);
+
+impl Future for SessionClosedFuture {
+	type Output = ();
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		match self.0.poll_next_unpin(cx) {
+			Poll::Pending => Poll::Pending,
+			// A message is only sent when
+			Poll::Ready(x) => {
+				tracing::info!("{:?}", x);
+				Poll::Ready(())
+			}
+		}
+	}
+}
+
+pub(crate) fn session_close() -> (SessionClose, SessionClosedFuture) {
+	// SessionClosedFuture is closed after one message has been recevied
+	// and max one message is handled then it's closed.
+	let (tx, rx) = tokio::sync::broadcast::channel(1);
```

Comment:
Yepp, I think the implementation is quite complicated/clever as the Receiver is not clone.

I think it just clone the message(s) to other receivers which is probably quite nice to avoid having a separate state for each receiver